### ISSUE TITLE
Multiple UV Layers

### ISSUE THREAD

**Issue by MarieRose1301** [powroupi/blender_mmd_tools#161]
Any ways of exporting multiple uv layers in a pmx file?

**Comment by powroupi** [comment]
Yea, simply add uv maps as needed via [UV Maps Panel](https://docs.blender.org/manual/en/dev/editors/uv_image/uv/layout_management.html#uv-maps-panel). However, there are some rules:
 - the first UV map will always be exported as base UV (UV0)
 - the UV map will be exported as xy channels of UVn (0< n <=4) if its name does not start with `_`, and a UV map will be exported as UVn's zw channels if the name eqauls to `_[UVn's name]`.
 - it is applied to each single object, so the same UV map name in different mesh objects may not be exported to be the same UVn

For example, if you created UV maps [ `baseUV`, `newUV`, `_testUV`, `testUV`, `_badUV` ] on object A, you get UV0.xy (`baseUV`), UV1.xy (`newUV`), UV2.xyzw (`testUV`+`_testUV`), `_badUV` will not be exported because `badUV` does not found. And if you created UV maps [ `bbUV`, `_testUV`, `testUV` ] on object B, you get UV0.xy (`bbUV`), UV1.xyzw (`testUV`+`_testUV`). If you export (object A + object B), the result will be combined.

**Comment by MarieRose1301** [comment]
I have a map model with 2 UV layers: DiffuseUV and LightmapUV, however when i get it exported to .pmx I really can't change to the lightmap UV layer - it still shows the diffuse UVs, here's the blend: https://sta.sh/01aiixyp64ii
I have been looking for a way to make the diffuse and lightmap textures blend in together in MMD but i don't really see the way to do so yet...

**Comment by powroupi** [comment]
MMD default shader use UV1 only for `SubTexture` mode of SPH/SPA textures, so I think you will need to write custom MME shaders to use addUVs in different way. Maybe @nathanvasil could help you. :)

And currently `mmd_tools` cannot display `SubTexture` mode of SPH/SPA textures properly yet.

**Comment by MarieRose1301** [comment]
oh thank you for the explanation c:

**Comment by MarieRose1301** [comment]
how can i contact @nathanvasil ?

**Comment by nathanvasil** [comment]
Using a subtexture slot will apply the subtexture, using AddUV 1, as a multiplication layer on top of your base texture.  For lightmaps, this is probably what you want: you're multiplying diffuse by light.

Another option is to bake your base texture from its UV onto your lightmap UV in order to work with a single image + UV.  For a heavily tiled texture, this might involve some loss of resolution; you should probably aim for a high resolution (like 4096^2) texture.  For info on this, see https://www.deviantart.com/vasilnatalie/art/Remapping-textures-from-alternate-UV-in-Blender-740710343 .

If you use either of these techniques, you'll probably want to use a diffuse 0 specular 0 ambient 1 material  to prevent any influence from the MMD light, as the texture won't really represent a straight diffuse/albedo value anymore, but have light baked into it already.

It's probably pointless to try to do this with shader work, since these other techniques will work as well and be more compatible, but if you want to play with methods of integrating addUV and other textures, autoluminous demonstrates how to get addUV coordinates from MMD/MME for use in shaders.  A lightmap shader is pretty basic, more basic than the default MMD shader, because all you're doing is getting a diffuse from one UV and multiplying it by light on a different UV-- no specular, nothing view dependent whatsoever.  Slightly fancier, all you'd be doing is adding this value to the output of the default MMD shader in order to combine with dynamic lighting (which would demand either lower light values or else some tone mapping, since MMD doesn't do any tone mapping by default, and your colors all get clamped to 0-1 range.)

**Comment by powroupi** [comment]
>I really can't change to the lightmap UV layer - it still shows the diffuse UVs, here's the blend: https://sta.sh/01aiixyp64ii

@MarieRose1301 you might try changing sphere map type of those meterials `['BLINN_S5R1Z_glass_', 'STAGE_SD018Z_amadoi', 'STAGE_SD018Z_glass_window', 'STAGE_SD018Z_hasira', 'STAGE_SD018Z_makuragi', 'STAGE_SD018Z_rail', 'STAGE_SD018Z_rail1']` from `Off` to `SubTexture`, then export to pmx to see if it works.

Here is the script that you can run in Blender Text Editor:
```python
import bpy
for m in bpy.data.materials:
    if m.texture_slots[2]:
        print(m.name)
        m.mmd_material.sphere_texture_type = '3' # SubTexture
```

**Comment by MarieRose1301** [comment]
@nathanvasil yeah basically, i just didn't know how to explain what i want correctly,sorry ^^"
i wouldn't want to bake my lightmaps into diffuse because the tiling is pretty big an almost all objects and it wouldn't really make much sense for me to do this, although i would have to resort to this if i can't find another way
[edit]
IT ALL WORKS NOW, THANK YOU <3

**Comment by MarieRose1301** [comment]
well it doesn't entirely... for some reason this happens
this is how it should look:

![image](https://user-images.githubusercontent.com/20154978/42619833-d37e79ee-85c1-11e8-86f9-296bd4ab4cc6.png)

this is what it looks like in pmx:

![image](https://user-images.githubusercontent.com/20154978/42620266-2efa786c-85c3-11e8-8389-64210e0f4aa7.png)

**Comment by powroupi** [comment]
Hmm, I don't find any texture slots that use LightmapUV, so it might not be the issue of UV layers, it's more like shading issue. I guess you need MME or custom shaders because the Normal Textures/Specular Textures/Reflection Textures do not seem to be suitable for MMD default shader. 

ps. Textures were not packed into your blend file, so I'm not really sure. :cry:

**Comment by MarieRose1301** [comment]
The textures weren't even assigned on import...
Here they are: https://sta.sh/019gdm6qvuay there's also the .fbx file with everything set as it should be

**Comment by nathanvasil** [comment]
Do you have it working?  The issue is only that you don't have the correct textures assigned.  On importing the .fbx, Blender doesn't know which textures go to which mat, at least, not in the rar you provided.  Here's an example of what I'm getting after following the procedure I described (ambient1, diffuse0, subtex lightmap) with my best guesses for diffuse + lightmap textures:

![image](https://user-images.githubusercontent.com/19930899/42644084-8c0be04a-85af-11e8-9afd-daaf78b0a43f.png)

**Comment by MarieRose1301** [comment]
@nathanvasil i have no idea what's wrong but i just cannot get it to show this way, it's all pitch black: 

![image](https://user-images.githubusercontent.com/20154978/42645729-76afecd4-8607-11e8-9118-b088f6c0ab3b.png)

**Comment by nathanvasil** [comment]
Keep working on it.  You now know it's possible.  Here are my recommendations:

1) If you don't speak Japanese, don't use the Japanese language PMXE.  Try https://www.deviantart.com/eoscustom3d/art/English-Pmx-Editor-v1-1-542100208 .

2) Rename your materials and textures with descriptive names in your native language so that you can easily keep track of them.

3) Delete everything but a single material to begin with.  Get that single material working.  Do the same thing with a second material and import it in.  One step at a time.  Reduce your system to the fewest number of parts possible for troubleshooting. don't try to troubleshoot everything at once.

If you have a .pmx file that you think should work but doesn't, and you've translated the relevant material names and texture names, and limited it to a single material, then host it someplace and I'll take a look.  (But I'm not going to dig through a million untranslated materials and texture files and deal with .fbx importation plus .pmx exportation.  That part's your job.)

**Comment by MarieRose1301** [comment]
it's not that i don't understand how to use pmx editor after 6 years of studying jp lol.... i'm just missing something. i have a pmx file with all textures applied: https://sta.sh/0dtj97r1ayr

**Comment by nathanvasil** [comment]
If you want me to look at it, simplify it to the smallest number of parts.  Like I wrote above.  In the file you've provided, there are 37 materials and who knows how many textures.  I should be looking at a file with 1 material and 2 textures.

**Comment by MarieRose1301** [comment]
a platform mesh, albedo and lightmap textures: https://sta.sh/023qf8pmeurb

**Comment by nathanvasil** [comment]
The .pmx you've provided does not have any addUV coordinates specified for its vertices.  There are a few ways you can check this: you can go to the vertex tab and take a look at some of the vertices-- notice that there is no add UV for any of these vertices (not that I looked at *all* of them....)  You can also enable the display of addUV on the 3D view (via the mode-> UVA -> UV1 button and menus, if you're working in English.)

I'm not sure why your verts are lacking addUV.  While I doubt that I am currently working with the most recent MMD Tools, I loaded the .fbx file you provided earlier, separated out the platform, created a new MMD model via shift-a->armature->New MMD model, parented the platform to the armature, and exported with default parameters, and addUV came along just fine.

**Comment by MarieRose1301** [comment]
just tried everything again, and it just won't work for me, so weird...
maybe i'm using a wrong version of blender/mmd tools or something?
tried a few times already and it really only exports the base UV layer

**Comment by nathanvasil** [comment]
You mean, that your exports don't have any addUV?  Replace your version of MMD Tools with the most recent, reset your Blender to factory defaults, disable any other non-standard add-ons (only ones that are standard are the import/export functions), and try again.

If it's still exporting without any addUV, describe the exact steps you're taking, including import and export parameters, upload the exact .fbx used (textures aren't really needed if it's just addUV) as well as the .pmx you get out.  Might be a bug with MMD Tools.

In the meantime, here's the model with valid addUV.  http://www.mediafire.com/file/rjwgxeegw9t0aa1/untitled.pmx/file

**Comment by MarieRose1301** [comment]
it's working now, not sure what caused it, but i downloaded blender 2.77 instead of 2.78 and reinstalled mmd_tools and the adduv layer is exported properly now
thank you for being patient with me

okay i have no idea why but the fbx converted with autodesk fbx converter worked while the noesis exported with new export argument one didn't :/

**Comment by MarieRose1301** [comment]
Sorry again but there's something strange happening
I've tried to export another map but for some reason subtex only works correctly with self-shadow OFF in mmd
any idea about what could be the cause, @nathanvasil ?
![image](https://user-images.githubusercontent.com/20154978/42703094-26d1da1a-86d4-11e8-8bb5-20aed2a9fcd8.png)
![image](https://user-images.githubusercontent.com/20154978/42703101-2a91944c-86d4-11e8-9396-cdba766c3019.png)

**Comment by nathanvasil** [comment]
Works fine for me. 

![image](https://user-images.githubusercontent.com/19930899/42710043-567a60da-8697-11e8-9c1b-7662ac1d758a.png)
![image](https://user-images.githubusercontent.com/19930899/42710079-78725d00-8697-11e8-9970-a5e08f01c05d.png)

**Comment by MarieRose1301** [comment]
the stage you converted works just fine for me too, but not this one... i'll try another one now

**Comment by MarieRose1301** [comment]
tried converting another one, it worked normally
weird haha

**Comment by MarieRose1301** [comment]
could you load it in mmd? it could be an issue with my pc or something: https://sta.sh/0samqcgojap

**Comment by nathanvasil** [comment]
No,  Work on issues yourself for a while (a week, I would say) before asking for help.  I'm willing to help every once in a while, but I've got my own things to do too.  If you put the work into it-- which includes simplifying the situation every time you have a difficult problem-- I think you'll find that you'll be able to solve most of your own problems.

**Comment by MarieRose1301** [comment]
hm i guess i cannot use both subtex and + sphere map type at the same time...

**Comment by powroupi** [comment]
>for some reason subtex only works correctly with self-shadow OFF in mmd

Interesting, I guess I found the answer. :)
It looks like SubTexture is using previous sphere texture's blend type (Multiply or Add). (only when using MME)

Assuming the sphere texture types of a object's materials are [`subTex1`, `sphMultiply`, `subTex2`, `sphAdd`, `subTex3`], then blend type of `subTex2` will be Multiply, blend type of `subTex3` will be Add, blend type of `subTex1` will be Multiply (default) at first, but it will change to Add after redrawing.

So, you can fix it by inserting a "sphMultiply" material after `STAGE_F9SD018Z_waku_002`. (remember to assign "sphMultiply" material to a face)

ps. sorry for my english, hope it helps. :blush:

**Comment by MarieRose1301** [comment]
@powroupi it doesn't work although i've thought it would make sense, but the thing is
if you have ANY material with environment map blending on + or x it will look like this with the self-shadow on. i have tested it by leaving only the lightmap materials on the pmx file which loaded just fine with self-shadow on. but loading any model with x or + type sphere maps causes the lightmaps to change the influence like this.

**Comment by powroupi** [comment]
Yeah, I did some tests and it works as expected. (not sure if it works with effects/shaders or not though)

I created 3 simple pmx files [`SphMultiply.pmx`, `SphAdd.pmx`, `SphSubTex.pmx`], each of them only contains 1 material/bone and simple mesh, then load them in order as following:
1. load `SphMultiply.pmx`, it's using Multiply as expected.
2. load `SphSubTex.pmx`, it's using Multiply as expected.
3. load `SphAdd.pmx`, it's using Add as expected.
4. load `SphSubTex.pmx`, it's using Add as expected while 2. is still using Multiply as expected.
5. load `SphMultiply.pmx`, it's using Multiply as expected.
6. load `SphSubTex.pmx`, it's using Multiply as expected while 4. is still using Add as expected.

So, we can control the blend type of SubTexture by reordering materials or inserting a dummy material. The pmx file you provided can be fixed by inserting a "SphMultiply" material after `STAGE_F9SD018Z_waku_002` and a "SphMultiply" material at beginning. :)
